File: app/src/test/java/com/mohsenoid/rickandmorty/ui/util/ImageDownloaderTest.java
package com.mohsenoid.rickandmorty.ui.util;

import com.mohsenoid.rickandmorty.data.service.network.NetworkHelper;
import com.mohsenoid.rickandmorty.executor.TaskExecutor;
import com.mohsenoid.rickandmorty.test.TestTaskExecutor;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;

public class ImageDownloaderTest {

    @Mock
    NetworkHelper networkHelper;

    private TaskExecutor testTaskExecutor = new TestTaskExecutor();
    private String cacheDirectoryPath = "";
    private ImageDownloader imageDownloader;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);

        imageDownloader = ImageDownloaderImpl.getInstance(networkHelper, cacheDirectoryPath, testTaskExecutor, testTaskExecutor);
    }

    @After
    public void tearDown() {
        ImageDownloaderImpl.instance = null;
    }

    @Test
    public void testExtractFileName() {
        // GIVEN
        String input = "https://rickandmortyapi.com/api/character/avatar/1.jpeg";
        String expected = "1.jpeg";

        // WHEN
        String actual = imageDownloader.extractFileName(input);

        // THAN
        assertEquals(expected, actual);
    }

    @Test
    public void testExtractNullFileName() {
        assertNull(imageDownloader.extractFileName(null));
    }
}

File: app/src/test/kotlin/com/mohsenoid/rickandmorty/view/util/ImageDownloaderTest.kt
package com.mohsenoid.rickandmorty.view.util

import com.mohsenoid.rickandmorty.data.network.NetworkHelper
import com.mohsenoid.rickandmorty.test.TestDispatcherProvider
import com.mohsenoid.rickandmorty.util.dispatcher.DispatcherProvider
import com.mohsenoid.rickandmorty.util.image.ImageDownloader
import com.mohsenoid.rickandmorty.util.image.ImageDownloaderImpl
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.mockito.Mock
import org.mockito.MockitoAnnotations

class ImageDownloaderTest {

    @Mock
    lateinit var networkHelper: NetworkHelper

    private val testDispatcherProvider: DispatcherProvider = TestDispatcherProvider()
    private val cacheDirectoryPath = ""

    private lateinit var imageDownloader: ImageDownloader

    @Before
    fun setUp() {
        MockitoAnnotations.initMocks(this)
        imageDownloader = ImageDownloaderImpl(
            networkHelper = networkHelper,
            cacheDirectoryPath = cacheDirectoryPath,
            dispatcherProvider = testDispatcherProvider
        )
    }

    @Test
    fun `test extractFileName`() {
        // GIVEN
        val input = "https://rickandmortyapi.com/api/character/avatar/1.jpeg"
        val expected = "1.jpeg"

        // WHEN
        val actual = imageDownloader.extractFileName(input)

        // THAN
        assertEquals(expected, actual)
    }
}
